core/src/main/scala/quasar/plugin/postgres/CsvSink.scala

```diff
@@ -0,0 +1,213 @@
+
+      case ProcessingFailure(q, _, e, p, t) =>
+        log.debug(s"PROCESSING_FAILURE: `$q` after ${(e + p).toMillis} ms (${e.toMillis} ms exec, ${p.toMillis} ms proc (failed)), detail: ${t.getMessage}", t)
+    }
+
+  private def error[F[_]: Sync](msg: => String, cause: => Throwable): F[Unit] =
+    Sync[F].delay(log.error(msg, cause))
+
+  private def debug[F[_]: Sync](msg: => String): F[Unit] =
+    Sync[F].delay(log.debug(msg))
+
+  private def trace[F[_]: Sync](msg: => String): F[Unit] =
+    Sync[F].delay(log.trace(msg))
+
+  private def logChunkSize[F[_]: Sync](c: Chunk[Byte]): F[Unit] =
+    trace[F](s"Sending ${c.size} bytes")
+
+  private def recordChunks[F[_]: Sync](total: Ref[F, Long])(c: Chunk[Byte]): F[Unit] =
+    total.update(_ + c.size) >> logChunkSize[F](c)
+
+  private def copyToTable(table: Table, columns: NonEmptyList[TableColumn]): Pipe[CopyManagerIO, Chunk[Byte], Unit] = {
+    val cols =
+      columns
+        .map(c => hygienicIdent(c.name))
+        .intercalate(", ")
+
+    val copyQuery =
+      s"COPY ${hygienicIdent(table)} ($cols) FROM STDIN WITH (FORMAT csv, HEADER FALSE, ENCODING 'UTF8')"
+
+    val logStart = debug[CopyManagerIO](s"BEGIN COPY: `${copyQuery}`")
+
+    val startCopy =
+      Stream.bracketCase(PFCM.copyIn(copyQuery) <* logStart) { (pgci, exitCase) =>
+        PFCM.embed(pgci, exitCase match {
+          case ExitCase.Completed => PFCI.endCopy.void
+          case _ => PFCI.isActive.ifM(PFCI.cancelCopy, PFCI.unit)
+        })
+      }
+
+    in => startCopy flatMap { pgci =>
+      in.map(_.toBytes) evalMap { bs =>
+        PFCM.embed(pgci, PFCI.writeToCopy(bs.values, bs.offset, bs.length))
+      }
+    }
+  }
+
+  private def createTable(table: Table, columns: NonEmptyList[TableColumn]): ConnectionIO[Int] = {
+    val preamble =
+      fr"CREATE TABLE" ++ Fragment.const(hygienicIdent(table))
+
+    val colSpecs =
+      columns
+        .map(c => Fragment.const(hygienicIdent(c.name)) ++ pgColumnType(c.tpe))
+        .intercalate(fr",")
+
+    (preamble ++ Fragments.parentheses(colSpecs))
+      .updateWithLogHandler(logHandler)
+      .run
+  }
+
+  private def dropTableIfExists(table: Table): ConnectionIO[Int] =
+    (fr"DROP TABLE IF EXISTS" ++ Fragment.const(hygienicIdent(table)))
+      .updateWithLogHandler(logHandler)
+      .run
+
+  private val pgColumnType: ColumnType.Scalar => Fragment = {
+    case ColumnType.Null => fr0"smallint"
+    case ColumnType.Boolean => fr0"boolean"
+    case ColumnType.LocalTime => fr0"time"
+    case ColumnType.OffsetTime => fr0"time with timezone"
+    case ColumnType.LocalDate => fr0"date"
+    // TODO: Can this be improved?
+    case ColumnType.OffsetDate => fr0"text"
+    case ColumnType.LocalDateTime => fr0"timestamp"
+    case ColumnType.OffsetDateTime => fr0"timestamp with time zone"
+    case ColumnType.Interval => fr0"interval"
+    case ColumnType.Number => fr0"numeric"
+    case ColumnType.String => fr0"text"
```

Comment:
I'm looking forward to when we have the type specification stuff. Importing everything as `text` and `numeric` is horrifying for indexing.

Reply:
Yeah, though one could refine the column types post-import and pre-indexing, for now.

Reply:
True